ODIN_II/SRC/netlist_statistic.cpp

```diff
@@ -0,0 +1,420 @@
+}
+
+// lower is better
+// we aim to approach 0
+double distance_to_goal(metric_t* golden_values, metric_t* calculated_values) {
+    return abs(circuit_effective_ratio(calculated_values) * circuit_effective_area(golden_values, calculated_values));
+}
+
+double fitness_calc (netlist_t* netlist, metric_t* m) {
+    long long node_count = netlist->total_net_count;
+    double area = circuit_area(m);
+
+    double area_efectness = (m->max_depth) / (area*node_count);
+    double fan_effectness = (m->max_fanin)+(m->max_fanout);
+    // printf("\n\n \t++node_count: %lld\n\t++area: %f\n\t++max_depth: %f\n\t++max_fan-in: %d\n\t++max_fan-out: %d\n\t++area_effectness: %f\n\t++fan_effectness: %f\n\n", 
+    //       node_count, area, m->max_depth, m->max_fanin, m->max_fanout, area_efectness, fan_effectness);
+
+
+    return 1000000 * (area_efectness) / (fan_effectness);
+}
+
+
+static void aggregate(metric_t* dest, metric_t** sources, int source_count, branching_type_e btype) {
+    int actual_count = 0;
+    init(dest);
+
+    // compute stats from parent
+    for (int i = 0; sources && i < source_count; i += 1) {
+        metric_t* src = sources[i];
+        if (src) {
+            actual_count += 1;
+            if (dest->min_depth == 0) {
+                dest->min_depth = src->min_depth;
+            } else {
+                dest->min_depth = std::min(src->min_depth, dest->min_depth);
+            }
+            dest->max_depth = std::max(src->max_depth, dest->max_depth);
+            dest->avg_depth += src->avg_depth;
+            dest->avg_width += src->avg_width;
+        }
+    }
+
+    if (actual_count) {
+        dest->avg_depth /= actual_count;
+        dest->avg_width /= actual_count;
+    }
+
+    if (btype == FANIN) {
+        for (int i = 0; sources && i < source_count; i += 1) {
+        metric_t* src = sources[i];
+        if (src) {
+            if (dest->min_fanin == 0) {
+                dest->min_fanin = src->min_fanin;
+            } else {
+                dest->min_fanin = std::min(src->min_fanin, dest->min_fanin);
+            }
+            dest->max_fanin = std::max(src->max_fanin, dest->max_fanin);
+        }
+    }
+    }
+    else if (btype == FANOUT) {
+        for (int i = 0; sources && i < source_count; i += 1) {
+            metric_t* src = sources[i];
+            if (src) {
+                if (dest->min_fanout == 0) {
+                    dest->min_fanout = src->min_fanout;
+                } else {
+                    dest->min_fanout = std::min(src->min_fanout, dest->min_fanout);
+                }
+                dest->max_fanout = std::max(src->max_fanout, dest->max_fanout);
+            }
+        }
+    }
+}
+
+static void add_net_to_stat(metric_t* dest, int branching_factor, branching_type_e btype) {
+    dest->min_depth += 1;
+    dest->max_depth += 1;
+    dest->avg_depth += 1;
+    dest->avg_width += branching_factor;
+    if (btype == FANIN) {
```

Comment:
this is redundant...
UPWARDS->min_width == fanin->min_width
DOWNWARD->max_width == fanout->max_width

remove that and everything related to it.

Reply:
what about max fanin? We need max fanin and fanout as criteria!